chain/network/src/snapshot_hosts/mod.rs

```diff
@@ -32,11 +36,143 @@ pub struct Config {
     /// least-recently-used cache. In the future, we may wish to
     /// implement something more sophisticated.
     pub snapshot_hosts_cache_size: u32,
+    /// The number of hosts we'll add to structures related to state part peer
+    /// selection each time we need to request parts from a new peer
+    pub part_selection_cache_batch_size: u32,
+}
+
+pub(crate) fn priority_score(peer_id: &PeerId, part_id: &PartId) -> [u8; 32] {
+    let mut h = Sha256::new();
+    h.update(peer_id.public_key().key_data());
+    h.update(part_id.idx.to_le_bytes());
+    h.finalize().into()
+}
+
+#[derive(Clone, Debug, PartialEq, Eq)]
+struct PartPriority {
+    score: [u8; 32],
+    // TODO: consider storing this on disk, so we can remember who hasn't
+    // been able to provide us with the parts across restarts
+    times_returned: usize,
+}
+
+impl PartPriority {
+    fn inc(&mut self) {
+        self.times_returned += 1;
+    }
+}
+
+impl PartialOrd for PartPriority {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl Ord for PartPriority {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        self.times_returned.cmp(&other.times_returned).then_with(|| self.score.cmp(&other.score))
+    }
+}
+
+#[derive(Default)]
+struct PartPeerSelector {
+    peers: BTreeMap<PartPriority, PeerId>,
```

Comment:
Incorporating the number of attempts into the comparator is an elegant choice, nicely done.

I believe you should be able to drop in a [BinaryHeap](https://doc.rust-lang.org/std/collections/struct.BinaryHeap.html#) instead for better performance. As far as I see you just need `pop` (get min element), `push` (insert), and `iter` (to get all peers). You don't seem to ever need access to arbitrary entries.

Reply:
another good point!